projects/driver_controls/src/x86/ads1015.c

```diff
@@ -1,28 +1,127 @@
+// This module emulates the behavior of ADS1015 on x86.
+// Instead of having interrupts raised by the device once a channel finishes conversion,
+// there is a soft timer with prv_timer_callback that is called periodically to update channel
+// readings and switch to the next channel. The rotation of the channels is implemented using
+// bitsets. The readings are just a arbitrary value.
 #include "ads1015.h"
-#include <status.h>
-#include <stdio.h>
+#include <string.h>
 #include "ads1015_def.h"
-#include "gpio_it.h"
+#include "soft_timer.h"
 
-// Initiates ads1015 by setting up registers and enabling ALRT/RDY Pin.
+#define ADS1015_CHANNEL_UPDATE_PERIOD 10
+#define ADS1015_CHANNEL_ARBITRARY_READING 0
+
+// Checks if a channel is enabled (true) or disabled (false).
+static bool prv_channel_is_enabled(Ads1015Storage *storage, Ads1015Channel channel) {
+  if (storage == NULL || channel >= NUM_ADS1015_CHANNELS) {
+    return false;
+  }
+  return ((storage->channel_bitset & (1 << channel)) != 0);
+}
+
+// Updates the channel_bitset when a channel is enabled/disabled.
+static void prv_mark_channel_enabled(Ads1015Channel channel, bool enable, uint8_t *channel_bitset) {
+  if (channel >= NUM_ADS1015_CHANNELS || channel_bitset == NULL) {
+    return;
+  }
+  if (enable) {
+    *channel_bitset |= (1 << channel);
+  } else {
+    *channel_bitset &= ~(1 << channel);
+  }
+}
+
+// Sets the current channel of the storage.
+static StatusCode prv_set_channel(Ads1015Storage *storage, Ads1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS || storage == NULL) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  storage->current_channel = channel;
+  return STATUS_CODE_OK;
+}
+
+// Periodically calls channels' callbacks imitating the interrupt behavior.
+static void prv_timer_callback(SoftTimerID id, void *context) {
+  if (context == NULL) {
```

Comment:
Nit: You only never call this with `context = storage` so why have a guard? If it is a `prv_` function you usually have total control over most of what you pass in so you don't need guard clauses. Same throughout the file. Usually you should sanitize before a callback?

Also since the address of `storage` will be retained in `context` even if `storage is gone` this check doesn't actually protect against anything in the event that `storage` becomes invalidated.

Reply:
I went through the private functions and had most of the guards removed!